src/eval_env.h

```diff
@@ -91,7 +172,7 @@ struct BindingEnv : public Env {
   const Rule* LookupRuleCurrentScope(const string& rule_name);
   const map<string, const Rule*>& GetRules() const;
 
-  void AddBinding(const string& key, const string& val);
+  void AddBinding(string&& key, string&& val);
```

Comment:
This is bad form.

If you want the function to take ownership of the string, pass it by value, not by rvalue-reference.

rvalue-reference parameters is only appropriate for "destructive" operations, and even that's debatable.

See further discussion here: https://github.com/redboltz/mqtt_cpp/issues/289#issuecomment-505193666

Reply:
> If you want the function to take ownership of the string, pass it by value, not by rvalue-reference.
> See further discussion here: redboltz/mqtt_cpp#289 (comment)

You're referring to item 41 (and 27?) from "Effective Modern C++"? But I want to accept rvalues only so that the compiler will also tell me when passing any lvalues as those will produce a copy. And passing by value would have an extra cost.

> rvalue-reference parameters is only appropriate for "destructive" operations, and even that's debatable.

your personal opinion?

Reply:
And please have a look at the code.

Reply:
I had to purchase a copy of the book to figure out what you were referring to.

Apparently Scott had the same idea I did.

Yes. I am referring to item 41 and 27.

What you are doing here is undesirable. It makes the code more complex. And doesn't actually solve and performance problems.


You've changed way too many things at once for me to be able to effectively review this pull request.

If you revert the changes to string parameters so that they require rvalue references, I would be happy to review the remaining code.

And no, that's not my opinion. Rvalue references are for "moving" an object, and moving is "destructive" (not destructor) to the original object. Perhaps my wording was weird, but my meaning wasn't derived from an opinion.